Add rbeesley's ansi-color.cmd, a CMD-based color tool (& more)

rbeesley opened the pull request:
Adds the Ansi-Color tool, completing #6470.

What started out as an experiment to see the support of ANSI support to
conhost, I wanted to write a Windows equivalent of Daniel Crisman's
[BASH script on tldp.org]. I didn't like how the BASH script hard coded
in whitespace and I thought Windows could do better. I applied
techniques to speed up the execution and tried to use the Command Script
batch language in ways that pushed the limits for what I thought it
could do. Running this from withing PowerShell, `&cmd /c ansi-color.cmd`,
I found out that the active code page is already 65001, but when ran
from a Command Prompt, the active code page depends on the regional
settings and I would have a dependency on CHCP to detect the active code
page, change to 65001 if necessary, and restore the previous code page
after running. I found it useful for designing color schemes and writing
shaders, and I started using an earlier version for logging bugs.

Initially it was a single script which I would dump every SGR
combination I could think of, and many which weren't implemented yet,
but as I was looking at other tools which had been written I wanted to
mimic the same output so I could do side-by-side comparisons. First I
wrote `crisman.def` and then `colortool.def`. So I had to align text.
Then output was slow for big tables so I wrote to an outbuffer and made
use of macros. I wanted to handle flags instead of needing to change
settings every time, so I added argument parsing and loading external
files.

It's a batch file that has some really useful purposes and does nothing
I've seen before. 

I've ran every definition in a Command Prompt, with CP 437 and CP 65001
to make sure it works. Posted as a Gist on my account for about year, a
Portuguese (Brazilian as I recall), speaking user tried the Gist and it
was failing because CHCP is localized. I've taken an approach to try and
make it work for different localizations, but this is a potential
problem. I also ran every definition in PowerShell 7, shelling down to
`cmd` to actually run it. Lastly, I went through using the flags and
made sure that help would be shown. Error messages generate error levels
when the script exits, so those could be used to use this in an
automated test and catch if there was a problem with the script
executing. It won't be able to validate if the generated output shows
correctly, but it would fail if a definition file is missing or if it
needed to switch to Unicode and wasn't flagged or configured to do so.
For trying to build the table stub and headers, there is some debug code
which will output what was parsed. This is the last debug code in the
script itself, but I found it to be useful at times, so there is a
configuration setting which can turn that debug output back on.
Technically there is also a debug statement to break after adding the
macros and parsing the arguments, but before it does any configuration
changes, changes the code page, or anything. This was useful for making
changes to macros and being able to test them as well as making sure
that flags and arguments are parsed correctly. It was a rather cryptic
discovery to call `cmd /c exit -1073741510` to break out, so in part
that was my reason to leave this in. The definition files themselves
could be cleaned up further, but in both of them there are a lot of
Unicode codepoints that could be useful when defining division lines for
the headers so I opted to just comment them out. I initially had the
default definition to require Unicode, but now it just changes to a
non-Unicode output if it can't. And this brings up the last concern. The
way certain settings are set, is that they are defined in the
`__TABLE__` section of the definition file.  With `PARSE_TABLE_DATA`,
the script looks line by line for `SET *` or `IF *` and if found it will
effectively eval that line. The definition files are written so that for
instance `SET "CELL= gYw "` could be used to set the code which is used.
`If` was allow so that there could be a test if Unicode were available
and set something different for ASCII and Unicode. But it also opens the
possibility that a rogue actor could create a `ansi-color.cmd`
definition file something like `SET "foo=bar" & DoSomethingBad.exe`.
First of all I'd be flattered that anyone would use this very niche
tool, but it is something I think which needs to be called out.
Essentially the definition files are just batch files, but they are
extremely limited to executing only one line at a time and only
supporting IF and SET commands. I think this is a minimal risk because
at that point batch files would already be more effective, but there
should also be an expectation that a batch file will run something and
the same may not be true for the heavily degraded definition files. I
think the risk is minimal but it is a risk I wanted to make clear. This
really showcases some interesting techniques and ideas, and I hope that
it is useful. All told, there are a couple years of incremental changes
in this PR.

Closes #6470

[BASH script on tldp.org]: https://tldp.org/HOWTO/Bash-Prompt-HOWTO/x329.html

zadjii-msft commented:
> tried to use the Command Script batch language in ways that pushed the limits for what I thought it could do

holy shit yeah it does. I've never seen a batch script anything like this before, but it's kinda incredible like that. You should probably throw your username at the top of this script in a cooment as the author, because you deserve the credit for whatever witchcraft this is. 

I'm cool accepting this as a tool to our repo, because it would be handy for debugging purposes, likely moreso than `colortool` itself. Lemme double-check that the rest of the team shares that consensus ☺️

Lo0oG commented on .github/actions/spelling/expect/205c1ec487b80f127ca611d10b973f66d9ed6812.txt:
```diff
@@ -0,0 +1,34 @@
+EINS
+EQU
+FGs
+forground
```
Shouldn't the spelling be fixed in src/tools/ansi-color/ansi-color.cmd rather than adding forground, Initalize, necessery, onscreen, properites & seperate to the expect lists?

zadjii-msft replied:
Yea, that'd probably be my main blocking comment here. It's also the holidays, so if we need to merge as is then I can quick fix this in post.

rbeesley replied:
@zadjii-msft , I can probably address this a little later today. I was also going to add the comment you suggested earlier too. Forground (sic) is intended because it is in the original comment, but I started fixing some of those other errors which are spelling errors.

zadjii-msft commented:
team consensus:
* this is a testament to what is possible with batch scripting
* no way in heck do any of us fully understand _how_ it works, so maintaining this would be likely impossible
* Would this be simpler as a C/C++ program? yea probably
* Will we accept it as is? Yea probably. It's a monument to batch scripting so it's pretty cool on that merit alone. Plus, it's a _tool_. It's not part of the CI, it's not something that ships, it's just a nice debug utility.

:shipit:

rbeesley commented:
It may read like Perl at times, but it should be complete now. If there is something you think is unclear, I can try to add more documentation, but I think the comments in the script should answer most questions and I tried to make the flow very modular. The main thing to remember is that to define the macros they need to be defined early on, but I wanted to add the configuration and data segment at the top. Configuration allows someone to set the condition of a flag, so they don't have to be passed on the command-line... actually that was the way this script was originally written, and the argument parsing was added later. Defining the CELL here is just a way to define something which doesn't have to be in definition files, and it actually makes it possible to use different definition files where just changing this one place changes whatever else is ran. I'm not sure any of the definitions I included use this technique, but you can imagine other ways this could be extended. The built-in definition is what I usually ran to review changes from build to build and when authoring color schemes. The main script doesn't start until about 362 lines in, but even this is designed to be very readable. Until 490, it is just trying to set the correct default values and settings, including changing the codepage if necessary. Then it reads the data segment, validates that the configuration supports the requirements of the definition, specifically UTF-8, then it resolves how the divisions should be drawn between table headings, cells, stubs, etc. Lastly it builds the table to a buffer and writes the buffer to the screen. After that it restores the codepage if it changed and exists. With labels and comments, I think it shouldn't be too difficult to read, but I've also rewritten it several times as I've introduced new functionality. As mentioned, it is a tool and a great demonstration that Command Scripting can accomplish more than expected.

rbeesley commented:
I'm happy with the current state. I made a few more changes to improve CHCP error handling and how things downgrade if that fails. Since the initial PR, I've also cleaned up the reading of the Data Segment so that it is more forgiving of case inconsistencies. There is more which could be done, but it seems to introduce performance problems to handle things like indentation in the definition files. It would be a nice to have, but probably not at that cost.

jsoref commented on src/tools/ansi-color/crisman.def:
```diff
@@ -0,0 +1,80 @@
+:: #
+:: #   This file echoes a bunch of color codes to the
+:: #   terminal to demonstrate what's available.  Each
+:: #   line is the color code of one forground color,
```
```suggestion
:: #   line is the color code of one foreground color,

```

jsoref commented:
Re:
> !CSI!2m

I recently added an ANSI color codes pattern to https://github.com/check-spelling/check-spelling/wiki/Configuration-Examples:-patterns:
```
# ANSI color codes
\\u001b\[\d+(?:;\d+|)m
```

You could probably do:
```
# ANSI color codes
!CSI!d+(?:;\d+|)m
```

rbeesley commented:
@jsoref, thanks for the feedback. I looked at the CSI pattern you suggested. I think it would actually need to be something [like](https://www.debuggex.com/r/mJHAFjf1D-rOkBsn): `(?i)!CSI!(?-i)(?:\d*[ABCDEFGJKST]|\d*;?\d*[HRf]|(?:\d+(?:;\d+|)+)?m|[45]i|6n)`, and I'm not sure that is going to suffice. It covers the following types of patterns as I've defined them in the script:

```
!CSI!m   # RESET
!csi!1m   # Intensity attribute with lower-case CSI variable... (this is cmd scripting so this is allowed)
!CSI!2;21;30m   # Faint, Double-Underscore, Black Foreground
!CSI!F   # Cursor Previous Line
```

And it supports all the CSI sequences I could quickly scrounge up. The problem I have with making this a pattern is that this isn't exhaustive. I know that it is an incomplete set of what could be [available](https://www.ecma-international.org/wp-content/uploads/ECMA-48_5th_edition_june_1991.pdf), and I don't know the set that Windows Terminal supports.

Something like `(?i)!CSI!(?-i)(?:(?:\d+(?:;\d+|)+)?[A-Za-z])` is a little more general, or even `(?i)!CSI!(?-i)(?:(?:;?\d|)+?[A-Za-z])` might be better, but very broad. Lastly, something like `![Cc][Ss][Ii]!(?:(?:;?\d|)+?[A-Za-z])` if the pattern checker wouldn't handle the case insensitivity flags, but I believe it is PCRE, so it should work.

The base minimum, following your example would be `!CSI!\d+(?:;\d+|)m`, you were missing a slash for the first digit.

Too much? Not enough? Using a pattern would probably be better but I'm not sure how deep to take this.

jsoref commented:
Sorry about the lost slash.

Personally I think that something is better than nothing, and it's an invitation for someone to improve it later. As long as you don't lose anything because of false positives, I see a pattern as a strict improvement.

Fwiw, Instead of using the fancy `(?-i)`, I'd just do this:
```
!(?:(?i)CSI)...
```

rbeesley commented:
@jsoref, nice. I didn't consider that the insensitivity could be applied in the non-capturing group, but of course it could.

```
!(?:(?i)CSI)!\d+(?:;\d+|)m
```

That's more readable than the form where you break it up into one character of groups. Final answer.

jsoref commented:
I should have clarified: the pattern matcher *is* Perl.

The comment includes customizable content:
https://github.com/microsoft/terminal/blob/main/.github/actions/spelling/advice.md

Here's my current default (which was written long after the above):
https://github.com/check-spelling/spell-check-this/blob/prerelease/.github/actions/spelling/advice.md

It looks like the `patterns` description in the advice could be clarified. If you're up for it, please feel free, otherwise maybe with my next upgrade PR, I'll see about adjusting it.

(I'm very close to a new [release](https://github.com/check-spelling/check-spelling/releases/tag/v0.0.20-alpha4), but still beating some [bugs out](https://github.com/check-spelling/check-spelling/compare/v0.0.20-alpha4..prerelease)....

DHowett reviewed (approved):
Ryan, this is a magnum opus and we're proud to have you contribute it to our humble repo. Thank you.
